authors/apps/friends/views.py

```diff
@@ -0,0 +1,135 @@
+        If same user, return an error message
+        Else delete the user relationship
+        Finally the serialized object of the unfollowed user is returned
+        """
+        try:
+            followed = get_user_model().objects.get(username=username)
+        except ObjectDoesNotExist:
+            raise Http404()
+
+        follower = get_user_model().objects.get(pk=request.user.id)
+
+        if followed == follower:
+            return Response(
+                {"message": "You cannot unfollow yourself"},
+                status=status.HTTP_406_NOT_ACCEPTABLE)
+
+        relationship = Friend.objects.filter(
+            user_from=follower, user_to=followed)
+
+        if not relationship:
+            return Response(
+                {"message": "you don't follow this user"},
+                status=status.HTTP_406_NOT_ACCEPTABLE)
+
+        relationship.delete()
+        user = get_user_model().objects.get(pk=followed.id)
+        serializer = FollowUnfollowSerializer(user)
+        user_id = serializer.data
+        if serializer:
+            return Response(user_id, status=status.HTTP_200_OK)
+
+
+class FollowersApiView(generics.ListAPIView):
+    """view for listing all followers of a particular user"""
+
+    def get_queryset(self):
+        """
+        check if the user associated with the username passed exists
+        if true return all the friend objects associated with the user
+        """
+        user = get_object_or_404(
+            get_user_model(), username=self.kwargs['username'])
+        return Friend.objects.select_related(
+            'user_from', 'user_to').filter(user_to=user.id).all()
+
+    def get(self, request, username, format=None):
+        """
+        Get & return the followers from the respective
+        object relationships i.e 'user_from'
+        """
+        friend_objects = self.get_queryset()
+        if friend_objects is not None:
+            followers = {u.user_from for u in friend_objects}
+            serializer = FollowersFollowingSerializer(followers, many=True)
+            return Response(serializer.data)
+
+
+class FollowingApiView(generics.ListAPIView):
+    """views to get all users a user is following"""
+
+    def get_queryset(self):
+        """
+        check if the user associated with the username passed exists
+        if true return all the friend objects associated with the user
+        """
+        user = get_object_or_404(
+            get_user_model(), username=self.kwargs['username'])
+        return Friend.objects.select_related(
+            'user_from', 'user_to', ).filter(user_from=user.id).all()
+
+    def get(self, request, username, format=None):
+        """
+        Get & return the 'followed users' from the
+        respective object relationships i.e 'user_to'
+        """
+        friend_objects = self.get_queryset()
+        if friend_objects is not None:
+            users_followed = {u.user_to for u in friend_objects}
```

Comment:
List comprehension is to create lists right? users_followed should be a list not dictionary

Reply:
Fixing that right away.